certbot/certbot/configuration.py

```diff
@@ -38,13 +59,17 @@ class NamespaceConfig:
     :ivar namespace: Namespace typically produced by
         :meth:`argparse.ArgumentParser.parse_args`.
     :type namespace: :class:`argparse.Namespace`
+    :ivar argument_sources: dictionary of argument names to their :class:`ArgumentSource`
+    :type argument_sources: :class:`Dict[str, ArgumentSource]`
 
     """
 
-    def __init__(self, namespace: argparse.Namespace) -> None:
+    def __init__(self, namespace: argparse.Namespace,
+                 argument_sources: Dict[str, ArgumentSource]) -> None:
```

Comment:
I think this is a breaking change to our public API.

I'm looking around in GitHub code search and I'm seeing a couple of projects that do construct `NamespaceConfig`, using only one argument to the constructor. I don't know if those projects are actively used or not, but it's probably good to stay on the safe side.

Is there a sensible default value that can work here?

Reply:
ahh shoot, thanks for catching this. well, the obvious default would just be to pass an empty dict as a default value, but this means that the `set_by_user(...)` check will always be false, which may break things in subtle ways for users.

i guess the safest option here would be to revert and wait for a major version release.

Reply:
Sooooooo we could maybe move `set_by_user` somewhere `_internal` instead of having it be a method on `NamespaceConfig`. Since `set_by_cli` was internal anyway and therefore people weren't (shouldn't have been) using it, so they probably don't need to be checking that anyway. Then we could set the default to `None` and use it more as a backwards compatibility flag and raise an error if we see it when we call `set_by_user`.

Reply:
I actually think this might make sense to do anyway, because I don't think people calling it will want to have to fix their usage of it especially when they're not actually using that part of the functionality, and this will let them continue to use more newer versions of certbot. In that case we could either add a default value and use it as a flag, or just use a different constructor/utility function internally that sets `argument_sources` after construction.

Reply:
so i agree that `set_by_user` shouldn't be in the public API, the problem is more that whenever we internally call it, a `NamespaceConfig` with an empty/None `argument_sources` will still always register all arguments as being not `set_by_user`. so e.g. https://github.com/certbot/certbot/blob/a5d223d1e55ce47067b64545bcecbf06bc53eb7e/certbot/certbot/_internal/storage.py#L285 will always be a nearly empty dict

Reply:
i guess i feel like proper usage of `NamespaceConfig` w/ this change will now always require setting up an accurate `argument_sources`, which is a somewhat tricky ask for public API

Reply:
What I'm trying to get at with all this is I think there's benefit to not breaking this interface at all instead of waiting for 3.0 given how it's used externally, and I think we can find a way to do that cleanly, whether it's one of these two things or something else.

Reply:
that makes sense! i'll open another PR that reverts the constructor interface & raises exceptions if `argument_sources` isn't set when `set_by_user` is called